### PR TITLE
Use nil object pattern for ethclient if disabled

### PR DESCRIPTION
**Pull request by j16r**

### CODE REVIEW

**Review by se3000 — DISMISSED** [review]

File: core/services/chainlink/application.go [hunk 1]
```diff
@@ -161,41 +161,27 @@ func (app *ChainlinkApplication) Start() error {
...
-	if ethEnabled {
-		err := app.Store.EthClient.Dial(context.TODO())
-		if err != nil {
-			return err
-		}
-	}
-
 	// XXX: Change to exit on first encountered error.
 	return multierr.Combine(
+		app.Store.EthClient.Dial(context.TODO()),
```
Comment (samsondav) [inline core/services/chainlink/application.go:166]: Interesting, should we have an application-wide context that gets "Done" on application close?

File: core/services/eth/client.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ type GethClient interface {
 	BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error)
 	FilterLogs(ctx context.Context, q ethereum.FilterQuery) ([]types.Log, error)
 	SubscribeFilterLogs(ctx context.Context, q ethereum.FilterQuery, ch chan<- types.Log) (ethereum.Subscription, error)
-	EstimateGas(ctx context.Context, call ethereum.CallMsg) (gas uint64, err error)
+	EstimateGas(ctx context.Context, call ethereum.CallMsg) (uint64, error)
```
Comment (samsondav) [inline core/services/eth/client.go:59]: Why this change?

File: core/services/eth/null_client.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+func (nc *NullClient) SubscribeFilterLogs(ctx context.Context, q ethereum.FilterQuery, ch chan<- types.Log) (ethereum.Subscription, error) {
+	return &nullSubscription{}, nil
+}
+
+func (nc *NullClient) SubscribeNewHead(ctx context.Context, ch chan<- *models.Head) (ethereum.Subscription, error) {
+	return &nullSubscription{}, nil
+}
+
+//
+// GethClient methods
```
Comment (samsondav) [inline core/services/eth/null_client.go:60]: Perhaps it should log something? or no?

File: core/store/store.go [hunk 4]
```diff
@@ -73,9 +73,14 @@ func newStoreWithKeyStore(
 		logger.Fatal(fmt.Sprintf("Unable to migrate key store to disk: %+v", e))
 	}
 
-	ethClient, err := eth.NewClient(config.EthereumURL(), config.EthereumSecondaryURL())
-	if err != nil {
-		logger.Fatal(fmt.Sprintf("Unable to create ETH client: %+v", err))
+	var ethClient eth.Client
+	if config.EthereumDisabled() {
+		ethClient = &eth.NullClient{}
```
Comment (samsondav) [inline core/store/store.go:78]: Probably we should log here

**Review by samsondav — APPROVED** [review]
Clean approach, I really like it. This is better than the old way imo

File: core/services/eth/client.go (hunk 2 above)
Comment (j16r) [reply core/services/eth/client.go:59]: Names in interfaces for arguments/returns are basically ignored / not useful.

File: core/services/chainlink/application.go (hunk 1 above)
Comment (j16r) [reply core/services/chainlink/application.go:166]: Not sure, this was just moved around.

**Review by se3000 — APPROVED** [review]
